File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/capturetag/CaptureTagEntity.kt
/*
 * CaptureTagEntity.kt
 *
 * Room @Entity for the `capture_tags` many-to-many join — links
 * captures to tags. Each row syncs independently (own id + dirty
 * + tombstone trio) so a tag added on phone A reaches phone B
 * without re-syncing the entire capture. See brief §2.
 *
 * Schema mirrors
 * `shared/design-system/migrations/quickink/v4_workspace.sql`.
 */

package app.quickink.mobile.data.capturetag

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import app.quickink.mobile.data.capture.CaptureEntity
import app.quickink.mobile.data.tag.TagEntity

@Entity(
    tableName = "capture_tags",
    foreignKeys = [
        ForeignKey(
            entity = CaptureEntity::class,
            parentColumns = ["id"],
            childColumns = ["capture_id"],
            // No cascade — capture_tags rows soft-delete
            // independently. A capture deletion tombstones the
            // capture; the join rows are cleaned up by the sync
            // worker separately so cross-device propagation works
            // even after the parent is gone locally.
            onDelete = ForeignKey.NO_ACTION,
        ),
        ForeignKey(
            entity = TagEntity::class,
            parentColumns = ["id"],
            childColumns = ["tag_id"],
            onDelete = ForeignKey.NO_ACTION,
        ),
    ],
    indices = [
        Index(value = ["capture_id"], name = "idx_capture_tags_capture"),
        Index(value = ["tag_id"], name = "idx_capture_tags_tag"),
        Index(value = ["dirty"], name = "idx_capture_tags_dirty"),
        Index(value = ["deleted_at"], name = "idx_capture_tags_tombstone"),
        // Logical unique-active constraint lives as a partial
        // index in SQL (see v4_workspace.sql); Room can't express
        // partial uniqueness so the plain index here just makes
        // joins fast. The app-side insert path is responsible for
        // dedup against tombstoned rows.
        Index(value = ["capture_id", "tag_id"], name = "idx_capture_tags_pair"),
    ],
)
data class CaptureTagEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "capture_id")
    val captureId: String,

    @ColumnInfo(name = "tag_id")
    val tagId: String,

    /**
     * Provenance of this tag assignment:
     *
     * - `"manual"` — user-typed in the tag picker bottom sheet.
     * - `"ai-suggested"` — written by the auto-tagging heuristic
     *   (Phase E). Kept as a flag so analytics can measure
     *   suggestion acceptance rate. Visually identical to manual
     *   tags after the user accepts (per brief §10 #9).
     * - `"migration"` — the v4 backfill row carrying the legacy
     *   `captures.category` value into the join. Lets us tell
     *   organic tagging activity apart from one-time migration
     *   noise in analytics dashboards.
     *
     * Stored as free-form TEXT so future sources land without a
     * migration.
     */
    @ColumnInfo(name = "source", defaultValue = "manual")
    val source: String = "manual",

    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean = true,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,
)

File: apps/quickink/ios/QuickInk/Data/CaptureTagEntity.swift
/*
 * CaptureTagEntity.swift
 *
 * GRDB record for the `capture_tags` many-to-many join — links
 * captures to tags. Each row syncs independently (own id + dirty
 * + tombstone trio) so a tag added on phone A reaches phone B
 * without re-syncing the entire capture.
 *
 * Schema mirrors
 * `shared/design-system/migrations/quickink/v4_workspace.sql`
 * + the iOS v8_workspace migration in `QuickInkDatabase.swift`.
 *
 * Mirror of `CaptureTagEntity.kt` in QuickInk's Android target.
 */

import Foundation
import GRDB

public struct CaptureTagEntity: Codable, FetchableRecord, PersistableRecord, Equatable, Sendable, Identifiable {

    public static let databaseTableName = "capture_tags"

    public var id: String
    public var captureId: String
    public var tagId: String
    /// Provenance of this tag assignment:
    ///   - `"manual"` — user typed/selected from the picker.
    ///   - `"ai-suggested"` — the auto-tagging heuristic (Phase E)
    ///     accepted by the user; surfaced visually identically to
    ///     manual once accepted but kept distinct here for analytics.
    ///   - `"migration"` — the v8 backfill row carrying the legacy
    ///     `captures.category` value into the join.
    public var source: String
    public var driveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public init(
        id: String,
        captureId: String,
        tagId: String,
        source: String = "manual",
        driveFileId: String? = nil,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id          = id
        self.captureId   = captureId
        self.tagId       = tagId
        self.source      = source
        self.driveFileId = driveFileId
        self.createdAt   = createdAt
        self.updatedAt   = updatedAt
        self.dirty       = dirty
        self.deletedAt   = deletedAt
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case captureId   = "capture_id"
        case tagId       = "tag_id"
        case source
        case driveFileId = "drive_file_id"
        case createdAt   = "created_at"
        case updatedAt   = "updated_at"
        case dirty
        case deletedAt   = "deleted_at"
    }
}

/// Projection for the per-tag document-count query — one row per
/// tag with active captures attached. Drives the home tag cloud
/// counts and the tag library cards' "N documents" subtitle.
public struct TagCount: Codable, FetchableRecord, Equatable, Sendable {
    public let tagId: String
    public let docCount: Int

    public enum CodingKeys: String, CodingKey {
        case tagId    = "tag_id"
        case docCount = "doc_count"
    }
}
